sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ComplexTypes.scala

```diff
@@ -39,7 +41,15 @@ object SimplifyExtractValueOps extends Rule[LogicalPlan] {
       // Remove redundant field extraction.
       case GetStructField(createNamedStruct: CreateNamedStruct, ordinal, _) =>
         createNamedStruct.valExprs(ordinal)
-
+      case GetStructField(w@WithFields(struct, names, valExprs), ordinal, maybeGetFieldName) =>
+        val getFieldName = maybeGetFieldName.getOrElse(w.dataType(ordinal).name)
+        val resolver = SQLConf.get.resolver
+        val matches = names.zip(valExprs).filter { case (name, _) => resolver(name, getFieldName) }
+        if (matches.nonEmpty) {
+          matches.last._2
```

Comment:
This is for case like `$"struct_col".withField("b", lit(100)).withField("b", lit(200))` while the last one takes effect. Can we add some comments to mention it?

Reply:
done.